cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+
+```JavaScript
+var o = {};
+Object.defineProperty(o, “a”, {
+    writable: true,
+    enumerable: true,
+    configurable: true,
+    value: “A”
+});
+```
+
+Apart from these, there are some special functions for object attributes. Object.preventExtensions() prevents new properties from being added to the object.
+
+## Handle errors in asynchronous calls
+
+Errors that occur within asynchronous callbacks can be easily forgettable. Therefore, as a general principle first argument to the asynchronous calls should be an Error object. Also, express routes handle errors itself, but it should be always remembered that errors occurred in asynchronous calls made within express routes are not handled, unless an Error object is sent as a first argument.
+
+Errors in these callbacks can be propagated as many times as possible. Each callback that the error has been propagated to can ignore, handle or propagate the error.
+
+## Listen to errors when using EventEmitter
+
+When using EventEmitter, errors can occur anywhere in the event chain. Normally, if an error occurs in an EventEmitter object, an error event with an Error object as its argument is called. However, if there are no attached listeners to that error event, the Error object that is sent as argument is thrown and becomes an uncaught exception. In short, if you do not handle errors within an EventEmitter object properly, these unhandled errors may crash your application. Therefore, you should always listen to error events when using EventEmitter objects.
+
+```JavaScript
+var events = require(‘events’);
+var myEventEmitter = function(){
+    events.EventEmitter.call(this);
+}
+require(‘util’).inherits(myEventEmitter, events.EventEmitter);
+myEventEmitter.prototype.someFunction = function(param1, param2) {
+    //in case of an error
+    this.emit(‘error’, err);
+}
+var emitter = new myEventEmitter();
+emitter.on(‘error’, function(err){
+    //Perform necessary error handling here
+});
+```
+
+## Handle uncaughtException
+
+Node.js behavior for uncaught exceptions is to print current stack trace and then terminate the thread. However, Node.js allows customization of this behavior. It provides a global object named process which is available to all Node.js applications. It is an EventEmitter object  and in case of an uncaught exception, “uncaughtException” event gets emitted and it is brought up to the main event loop. In order to provide a custom behavior for uncaught exceptions, you can bind to this event. However, resuming the application after such an uncaught exception can lead to further problems. Therefore, if you do not want to miss any uncaught exception, you should bind to uncaughtException event and cleanup any allocated resources like file descriptors, handles and similar before shutting down the process. Resuming the application is strongly discouraged as the application will be in an unknown state.
```

Comment:
Also good not to print a stacktrace to the users - as it leaks information

Reply:
added the recommendation